polynote-kernel/src/main/scala/polynote/kernel/PolyKernel.scala

```diff
@@ -31,70 +31,58 @@ class PolyKernel private[kernel] (
     */
   protected lazy val taskQueue: TaskQueue = TaskQueue(statusUpdates).unsafeRunSync()
 
-  private def runPredef(kernel: LanguageKernel[IO], language: String): IO[Unit] =
-    kernel.predefCode.fold(kernel.init()) {
-      code => for {
-        oq <- Queue.unbounded[IO, Result]
-        _  <- taskQueue.runTaskIO(s"Predef $language", s"Predef ($language)")(_ => kernel.runCode("Predef", Nil, Nil, code, oq, statusUpdates))
-      } yield ()
-  }
-
-  protected def getKernel(language: String): IO[LanguageKernel[IO]] = Option(kernels.get(language)).map(IO.pure).getOrElse {
+  protected def getKernel(language: String): IO[LanguageKernel[IO, GlobalInfo]] = Option(kernels.get(language)).map(IO.pure).getOrElse {
     launchingKernel.acquire.bracket { _ =>
       Option(kernels.get(language)).map(IO.pure).getOrElse {
         for {
           factory <- IO.fromEither(Either.fromOption(subKernels.get(language), new RuntimeException(s"No kernel for language $language")))
-          kernel  <- taskQueue.runTask(s"Kernel$$$language", s"Starting $language kernel")(_ => factory.apply(dependencies.getOrElse(language, Nil), symbolTable))
+          kernel  <- taskQueue.runTask(s"Kernel$$$language", s"Starting $language kernel")(_ => factory.apply(dependencies.getOrElse(language, Nil), globalInfo))
           _        = kernels.put(language, kernel)
-          _       <- runPredef(kernel, language)
         } yield kernel
       }
     } {
       _ => launchingKernel.release
     }
   }
 
-  protected def withKernel[A](cellId: String)(fn: (Notebook, NotebookCell, LanguageKernel[IO]) => A): IO[A] = for {
+  protected def withKernel[A](cellId: String)(fn: (Notebook, NotebookCell, LanguageKernel[IO, GlobalInfo]) => A): IO[A] = for {
     notebook <- getNotebook()
     cell     <- IO.fromEither(Either.fromOption(notebook.cells.find(_.id == cellId), new NoSuchElementException(s"Cell $cellId does not exist")))
     kernel   <- getKernel(cell.language)
   } yield fn(notebook, cell, kernel)
 
-  private def prevCells(notebook: Notebook, id: String) = "Predef" :: notebook.cells.view.takeWhile(_.id != id).map(_.id).toList
+  private def parentCellId(notebook: Notebook, id: String) = notebook.cells.view.takeWhile(_.id != id).map(_.id.toString).head
+
+  private def parentContext(notebook: Notebook, id: String) = Option(runtimeContextHolder.get(parentCellId(notebook, id))).getOrElse(throw new Exception("Well, how did this happen..."))
 
-  protected def findAvailableSymbols(prevCells: List[String], kernel: LanguageKernel[IO]): Seq[kernel.Decl] = {
-    symbolTable.currentTerms.filter(v => prevCells.contains(v.sourceCellId)).asInstanceOf[Seq[kernel.Decl]]
+  private val runtimeContextHolder: ConcurrentHashMap[String, RuntimeContext[GlobalInfo]] = new ConcurrentHashMap[String, RuntimeContext[GlobalInfo]] {
+    val predef = RuntimeContext.getPredefContext[globalInfo.type](globalInfo)
+    predef.cellId -> predef
```

Comment:
I realized after simplifying RuntimeContext that there was still a need to look up the proper RuntimeContext, so I quickly brought the Map back here. 

One alternative would be to have both prev and next pointers inside RuntimeContext, so that we only need to initialize the predef context and just update the nodes as needed - a true doubly linkedlist like we discussed originally. I haven't thought too much about it but one worry I have is whether we'd need to synchronize around any modifications to the list - I'm a bit concerned that could get a bit gnarly (we might need a var to hold it, etc). 

I don't yet have a story for how to update the nodes if cells get inserted - I think that if you insert two cells in a row (or run a cell without running the one before) things will barf (as there'd be no entry for the parent cell Id)...